File: JanRotation/KalmanModel.jl
using LinearAlgebra
using StatsFuns
using Distributions

function getKalman(A::Array{Float64,2}, C::Array{Float64,2}, totalI::Int, initVar::Array{Float64,2})
    if size(initVar,1)!=size(initVar,2)
        initVar = diagm(initVar)
    end
    Vars = Array{Array{Float64,2}, 1}(undef, totalI)
    Ks = Array{Array{Float64,2}, 1}(undef, totalI)
    Vars[1] = initVar
    #Get filter
    for i = 2:totalI
        Ks[i] = (A*Vars[i-1]*A'*C')/(C*A*Vars[i-1]*A'*C' .+ C*initVar*C')
        Vars[i] = (I - Ks[i]*C)*A*Vars[i-1]*A'
    end
    return Ks, Vars
end

struct KalmanModel{A2<:Array{Float64,2}, F<:Float64, I<:Int64, A1<:Array{Float64,1}, AA<:Array{Array{Float64,2},1}, FN<:Function}
    A::A2
    C::A2
    muPrior::A1
    initVar::A2
    a::F
    totalI::I
    Ks::AA
    Vars::AA
    runSim::FN
end


function runSim(Zs::Array{Array{Float64,1},1}, Ys::Array{Array{Float64,1},1}, KM::KalmanModel)
    Mus = Array{Float64,2}(undef, length(Zs[1])+1, KM.totalI)
    Mus[:,1] = vcat(KM.muPrior, 0)
    for i = 2:KM.totalI
        sum!(Mus[1:end-1,i], KM.A*Mus[1:end-1,i-1] + KM.Ks[i]*(Ys[i] - KM.C*KM.A*Mus[1:end-1,i-1]))
        Mus[end,i] = logaddexp(Mus[end,i-1], KM.getLLike(KM.C*Mus[1:end-1,i], KM.C*KM.Vars[i]*KM.C', Ys[i]))
    end
    Mus[1:end-1,:] -= hcat(Zs...)
    return Mus
end

function runSim(Mus::Array{Float64,2}, totalI::Int, A::Array{Float64,2}, C::Array{Float64,2},
                Ks::Array{Array{Float64,2},1}, Vars::Array{Array{Float64,2},1}, LLike::Function,
                Zs::Array{Array{Float64,1},1}, Ys::Array{Array{Float64,1},1})
    for i = 2:totalI
        sum!(Mus[1:end-1,i], A*Mus[1:end-1,i-1] + Ks[i]*(Ys[i] - C*A*Mus[1:end-1,i-1]))
        # Mus[end,i] = LLikeSum(Mus[end,i-1], C*Mus[1:end-1,i], C*Vars[i]*C', Ys[i])
        Mus[end,i] = LLike(C*Mus[1:end-1,i], C*Vars[i]*C', Ys[i])
    end
    Mus[1:end-1,:] -= hcat(Zs...)
    # cumsum!(Mus[end,:], Mus[end,:])
    accumulate!(logaddexp, Mus[end,:], Mus[end,:])
    return Mus
end

function getRunSim(A, C, muPrior, initVar, a, totalI, Ks, Vars)
    Mus = Array{Float64,2}(undef, length(muPrior)+1, totalI)
    Mus[:,1] = vcat(muPrior, 0)
    nCon = ((2*pi)^(-length(muPrior)/2))
    # LLikeSum = (last, Mu, Cov, Y)->logaddexp(last, log(nCon*(det(Cov)^(-1/2))*exp(-((Y-Mu)'*inv(Cov)*(Y-Mu))/2)))
    LLike = (Mu, Cov, Y)->log(nCon*(det(Cov)^(-1/2))*exp(-((Y-Mu)'*inv(Cov)*(Y-Mu))/2))
    return (Zs::Array{Array{Float64,1},1}, Ys::Array{Array{Float64,1},1}) -> runSim(Mus, totalI, A, C, Ks, Vars, LLike, Zs, Ys)
end


KalmanModel(A::Array{Float64,2}, C::Array{Float64,2}, muPrior::Array{Float64,1}, initVar::Array{Float64,1}, a::Float64, totalI::Int) =
    KalmanModel(A, C, muPrior, diagm(initVar), a, totalI, getKalman(A, C, totalI, diagm(initVar))...)
KalmanModel(A::Array{Float64,2}, C::Array{Float64,2}, muPrior::Array{Float64,1}, initVar::Array{Float64,2}, a::Float64, totalI::Int, Ks::Array{Array{Float64,2},1}, Vars::Array{Array{Float64,2},1}) =
    KalmanModel(A, C, muPrior, initVar, a, totalI, Ks, Vars, getRunSim(A, C, muPrior, initVar, a, totalI, Ks, Vars))

File: JanRotation/KalmanModel.m
classdef KalmanModel < handle
    properties
        A
        C
        muPrior
        initVar
        a
        totalI
        Ks
        Vars
        blankMus
    end
    
    methods
        function obj = KalmanModel(varargin)
            props = properties(obj);
            for v = varargin
                obj.(props{1}) = v{1};
                props = props(2:end);
            end
            
            if size(obj.initVar,2) ~= size(obj.initVar,1)
               obj.initVar = diag(obj.initVar);
            end
            obj.Vars = NaN([size(obj.initVar), obj.totalI]);
            obj.Ks = NaN([fliplr(size(obj.C)), obj.totalI]);
            obj.Vars(:,:, 1) = obj.initVar;
%           Get filter
            for i = 2:obj.totalI
                obj.Ks(:,:,i) = (obj.A*squeeze(obj.Vars(:,:,i-1))*obj.A'*obj.C')/(obj.C*obj.A*obj.Vars(:,:,i-1)*obj.A'*obj.C' + obj.C*obj.initVar*obj.C');
                obj.Vars(:,:,i) = (eye(size(obj.Ks(:,:,i)*obj.C,1)) - obj.Ks(:,:,i)*obj.C)*obj.A*obj.Vars(:,:,i-1)*obj.A';
            end
            
            obj.blankMus = NaN(length(obj.muPrior)+1, obj.totalI);
        end
        
        function [SEs, eYs, Mus] = runSim(obj, Zs, Ys)
%             Like = @(Y, Mu, Cov) ((2*pi)^(-length(obj.muPrior)/2))/(sqrt(det(Cov))) * expm1(-((Y-Mu)'*(Cov\(Y-Mu)))/2);
%             LLike = @(Y, Mu, Cov) log1p(((2*pi)^(-length(obj.muPrior)/2))/(sqrt(det(Cov)))) + (-((Y-Mu)'*(Cov\(Y-Mu)))/2);%using log1p and expm1 are hackss
            LLike = @(Y, Mu, Cov) -log1p(sqrt(det(Cov)*(2*pi)^size(obj.C, 1))) + (-((Y-Mu)'*(Cov\(Y-Mu)))/2);%using log1p and expm1 are hackss
            Mus = obj.blankMus;
            Mus(:, 1) = [obj.muPrior; 0];
            for i = 2:obj.totalI
                Mus(1:end-1,i) = obj.A*Mus(1:end-1,i-1) + obj.Ks(:,:,i)*(Ys(:,i) - obj.C*obj.A*Mus(1:end-1,i-1));
%                 Mus(end,i) = logmulexp(Mus(end,i-1), LLike(Ys(:,i), obj.C*Mus(1:end-1,i), obj.C*obj.Vars(:,:,i)*obj.C'));
                Mus(end,i) = obj.logsumexp([Mus(end,i-1), log(LLike(Ys(:,i), obj.C*Mus(1:end-1,i), obj.C*obj.Vars(:,:,i)*obj.C'))], 2);
%                 Mus(end,i) = LLike(Ys(:,i), obj.C*Mus(1:end-1,i), obj.C*obj.Vars(:,:,i)*obj.C');
            end
%             Mus(end,:) = cumsum(Mus(end,:));
            SEs = [(real(Mus(1:end-1,:)) - Zs).^2; Mus(end,:)];
            eYs = obj.C*real(Mus(1:end-1,:));
            return
        end
        
        function s = logcumsumexp(~, x, w, dim)
            % Compute log(sum(exp(x),dim)) while avoiding numerical underflow.
            switch nargin 
                case 2
                    % Determine which dimension sum will use
                    dim = find(size(x)~=1,1);
                    if isempty(dim), dim = 1; end
                    w = ones(1, size(x,dim));
                    
                case 3
                    if isscalar(w)
                        dim = w;
                        w = ones(1, size(x,dim));
                    else
                        dim = find(size(x)~=1,1);
                        if isempty(dim), dim = 1; end
                    end
                    
                case 4
                    if isempty(w)
                        w = ones(1, size(x,dim));
                    end
            end

            % subtract the largest in each dim
            y = max(x,[],dim);
            x = bsxfun(@minus,x,y);
            s = y + log(cumsum(w.*exp(x),dim));
            i = find(~isfinite(y));
            if ~isempty(i)
                s(i) = y(i);
            end
            return
        end
        
        function s = logsumexp(~, x, w, dim)
            % Compute log(sum(exp(x),dim)) while avoiding numerical underflow.
            switch nargin 
                case 2
                    % Determine which dimension sum will use
                    dim = find(size(x)~=1,1);
                    if isempty(dim), dim = 1; end
                    w = ones(1, size(x,dim));
                    
                case 3
                    if isscalar(w)
                        dim = w;
                        w = ones(1, size(x,dim));
                    else
                        dim = find(size(x)~=1,1);
                        if isempty(dim), dim = 1; end
                    end
                    
                case 4
                    if isempty(w)
                        w = ones(1, size(x,dim));
                    end
            end

            % subtract the largest in each dim
            y = max(x,[],dim);
            x = bsxfun(@minus,x,y);
            s = y + log(sum(w.*exp(x),dim));
            i = find(~isfinite(y));
            if ~isempty(i)
                s(i) = y(i);
            end
            return
        end
        
    end
    
    
end